File: Data Structures & Algorithms/buy-and-sell-crypto/submission-1.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        l, r = 0, 1
        sol = 0

        while r <= len(prices)-1:
            #print(f"l: {l}, r: {r}")
            if prices[l] > prices[r]:
                l = r
                r += 1
                #print(f"l: {l}, r: {r}")
                continue;

            profit = prices[r]-prices[l]
            sol = max(sol, profit)
            #print(profit, prices[l], prices[r])
            r += 1
        return sol

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        unordered_map<char, int> counts;
        int max_count = 0;
        int sol = 0;
        int l = 0, r = 0;

        while (r < s.size()) {
            counts[s[r]]++;
            max_count = max(max_count, counts[s[r]]);

            if ((r-l+1) - max_count > k) {
                counts[s[l]]--;
                l++;
            }

            sol = max(sol, r-l+1);
            r++;
        }
        return sol;
    }
};
